### PR TITLE
fix(grpc): linking multiple proto files with the same package name

### PR DESCRIPTION
**Pull request by meskill**
**Summary:**  
Instead of searching for the specific FileDescriptorSet based on package name, collect all of the linked files into single FileDescriptorSet that will be used for searching services and messages

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

## Summary by CodeRabbit

- **New Features**
	- Enhanced error messaging for missing protobuf files in gRPC configuration.
	- Introduced a scenario test for handling multiple Protocol Buffer files with the same package name.

- **Bug Fixes**
	- Updated logic for constructing service names in gRPC methods to ensure accuracy.

- **Refactor**
	- Simplified the handling of file descriptor sets in gRPC configurations.
	- Improved merging logic for gRPC extensions.

- **Tests**
	- Added unit tests for new merging behaviors in gRPC configurations.

### CODE REVIEW

File: src/blueprint/operators/grpc.rs [hunk 1]
```diff
@@ -166,8 +166,8 @@ pub fn compile_grpc(inputs: CompileGrpc) -> Valid<Expression, String> {
     Valid::from(GrpcMethod::try_from(grpc.method.as_str()))
         .and_then(|method| {
             Valid::from_option(
-                config_module.extensions.get_file_descriptor_set(&method),
-                format!("File descriptor not found for method: {}", grpc.method),
+                config_module.extensions.get_file_descriptor_set(),
```
Comment (tusharmath) [inline src/blueprint/operators/grpc.rs:169]: We need to validate if the path to the method exists in the proto file.

File: src/config/config_module.rs [hunk 2]
```diff
@@ -153,3 +163,93 @@ impl From<Config> for ConfigModule {
         ConfigModule { config, input_types, output_types, ..Default::default() }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    mod extensions {
+        mod merge_right {
```
Comment (tusharmath) [inline src/config/config_module.rs:170]: Do we need these tests for merge_right? We are using macros now, so we should be able to depend on them.

File: tests/snapshots/execution_spec__test-grpc-invalid-proto-id.md_errors.snap [hunk 3]
```diff
@@ -4,7 +4,7 @@ expression: errors
 ---
 [
   {
-    "message": "File descriptor not found for method: abc.NewsService.GetAllNews",
+    "message": "Protobuf files were not specified in the config",
```
Comment (tusharmath) [inline tests/snapshots/execution_spec__test-grpc-invalid-proto-id.md_errors.snap:7]: I think this should be validated at compile time.

File: src/blueprint/operators/grpc.rs (hunk 1 above)
Comment (meskill) [reply src/blueprint/operators/grpc.rs:169]: we have multiple proto files and we'll check that method exists later when executing method resolving

File: src/config/config_module.rs (hunk 2 above)
Comment (meskill) [reply src/config/config_module.rs:170]: I've added impl MergeRight for external type FileDescriptorSet

File: tests/snapshots/execution_spec__test-grpc-invalid-proto-id.md_errors.snap (hunk 3 above)
Comment (meskill) [reply tests/snapshots/execution_spec__test-grpc-invalid-proto-id.md_errors.snap:7]: this is validation error snapshot for compilation stage

File: tests/execution/grpc-proto-with-same-package.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    repeated News news = 1;
+}
+```
+
+```graphql @server
+schema
+  @server(port: 8000, graphiql: true)
+  @upstream(httpCache: true, batch: {delay: 10})
+  @link(id: "errors", src: "errors.proto", type: Protobuf)
+  @link(id: "news", src: "news.proto", type: Protobuf) {
```
[the remark this answers is not in the document]
Comment (meskill) [reply tests/execution/grpc-proto-with-same-package.md:55]: actually, copy-pasted from another test and kept the same approach

File: src/blueprint/operators/grpc.rs (hunk 1 above)
Comment (tusharmath) [reply src/blueprint/operators/grpc.rs:169]: Understood, but we are dropping a significant feature here, I don't see why we need to move that check to runtime when we can identify the issue at compile time.

File: tests/execution/grpc-proto-with-same-package.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tusharmath) [reply tests/execution/grpc-proto-with-same-package.md:55]: Let's add a simpler test here: Use types such as Foo and Bar in the same package. Also, we should test by making requests to the services defined in the two files. The current implementation has the error proto but it's not being used anywhere.

File: src/blueprint/operators/grpc.rs (hunk 1 above)
Comment (meskill) [reply src/blueprint/operators/grpc.rs:169]: The check it later in this file when calling `to_operation` and it's at compile time.

The error looks like this https://github.com/tailcallhq/tailcall/blob/2ddb53663be9b6204187076ce286bb56a0d8df71/tests/snapshots/execution_spec__test-grpc-service.md_errors.snap#L7

**Review by tusharmath — APPROVED** [review]

**Pull request merged**